opt: correctly type lookup join constant projections

mgartner opened the pull request:
#### opt/lookupjoin: show type of projected constants in tests

Release note: None

#### opt: correctly type lookup join constant projections

Fixes #134697

Release note (bug fix): A bug has been fixed that could cause incorrect
query results when the optimizer planned a lookup join on an index
containing a column of type `CHAR(N)`, `VARCHAR(N)`, `BIT(N)`,
`VARBIT(N)`, or `DECIMAL(M, N)`, and the query held that column constant
to a single value (e.g., with an equality filter).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134891)
<!-- Reviewable:end -->

mgartner committed:
opt/lookupjoin: show type of projected constants in tests

Release note: None

mgartner committed:
opt: correctly type lookup join constant projections

Fixes #134697

Release note (bug fix): A bug has been fixed that could cause incorrect
query results when the optimizer planned a lookup join on an index
containing a column of type `CHAR(N)`, `VARCHAR(N)`, `BIT(N)`,
`VARBIT(N)`, or `DECIMAL(M, N)`, and the query held that column constant
to a single value (e.g., with an equality filter).

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice fix!

> Reviewed 7 of 7 files at r2, 3 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134891#-:-OBXRGXNEiJCq1J1xKMJ:b-46lt7k)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/lookupjoin/constraint_builder.go` line 316 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/134891#-OBXR4G50HsnQM7TXVDP:-OBXR4G50HsnQM7TXVDQ:b7ead55) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b757766ae96cb09c3c37b2c7d160792eb5dd237/pkg/sql/opt/lookupjoin/constraint_builder.go#L316)):*
> ```Go
> 		// and use it as an equality column.
> 		if ok && len(foundVals) == 1 {
> 			idxColType := foundVals[0].ResolvedType()
> ```

nano-nit: might be worth giving this variable a different name, like `constColType` or something


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134891#-:-OBXRzzP4bSbui4gEbyR:b-oduykh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/opt/lookupjoin/constraint_builder.go` line 316 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/134891#-OBXR4G50HsnQM7TXVDP:-OBXRxdBCXECA-dSa6dD:b-s516mu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b757766ae96cb09c3c37b2c7d160792eb5dd237/pkg/sql/opt/lookupjoin/constraint_builder.go#L316)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nano-nit: might be worth giving this variable a different name, like `constColType` or something
</blockquote></details>

Good catch. I simplified to just `typ`.


<!-- Sent from Reviewable.io -->
